File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length;
        int curMax = nums[0], curMin = nums[0];
        int maxProd = curMax;
        for(int i=1;i<n;i++){
            int temp =curMax * nums[i];
            curMax = Math.max(nums[i], Math.max(temp, nums[i]*curMin));
            curMin = Math.min(nums[i], Math.min(temp, nums[i]*curMin));
            maxProd = Math.max(maxProd, curMax);
        }
        return maxProd;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        //idea is to maintain max and min products 
        //such that it handles negative procuts
        int curMax=nums[0],curMin=nums[0],maxProd=curMax;
        int n=nums.size();
        for(int i=1;i<n;i++){
            int temp = curMax;
            curMax = max(nums[i],max(curMax*nums[i],curMin*nums[i]));
            curMin = min(nums[i],min(temp*nums[i],curMin*nums[i]));
            maxProd=max(curMax,maxProd);
        }
        return maxProd;
    }
};